app/src/main/java/com/duckduckgo/widget/FavoritesWidgetService.kt

```diff
@@ -0,0 +1,174 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.widget
+
+import android.appwidget.AppWidgetManager
+import android.content.Context
+import android.content.Intent
+import android.graphics.Bitmap
+import android.os.Bundle
+import android.widget.RemoteViews
+import android.widget.RemoteViewsService
+import androidx.core.graphics.drawable.toBitmap
+import androidx.core.net.toUri
+import com.duckduckgo.app.bookmarks.model.FavoritesRepository
+import com.duckduckgo.app.browser.BrowserActivity
+import com.duckduckgo.app.browser.R
+import com.duckduckgo.app.browser.favicon.FaviconManager
+import com.duckduckgo.app.global.DuckDuckGoApplication
+import com.duckduckgo.app.global.domain
+import com.duckduckgo.app.global.view.generateDefaultDrawable
+import kotlinx.coroutines.runBlocking
+import javax.inject.Inject
+
+class FavoritesWidgetService : RemoteViewsService() {
+
+    companion object {
+        const val MAX_ITEMS_EXTRAS = "MAX_ITEMS_EXTRAS"
+        const val THEME_EXTRAS = "THEME_EXTRAS"
+    }
+
+    override fun onGetViewFactory(intent: Intent): RemoteViewsFactory {
+        return FavoritesWidgetItemFactory(this.applicationContext, intent)
+    }
+
+    class FavoritesWidgetItemFactory(val context: Context, intent: Intent) : RemoteViewsFactory {
+
+        private val theme = WidgetTheme.getThemeFrom(intent.extras?.getString(THEME_EXTRAS))
+
+        @Inject
+        lateinit var favoritesDataRepository: FavoritesRepository
+
+        @Inject
+        lateinit var faviconManager: FaviconManager
+
+        @Inject
+        lateinit var widgetPrefs: WidgetPreferences
+
+        private val appWidgetId = intent.getIntExtra(
+            AppWidgetManager.EXTRA_APPWIDGET_ID,
+            AppWidgetManager.INVALID_APPWIDGET_ID
+        )
+
+        private val faviconItemSize = context.resources.getDimension(R.dimen.savedSiteGridItemFavicon).toInt()
+        private val faviconItemRadius = context.resources.getDimension(R.dimen.savedSiteGridItemRadiusFavicon).toInt()
+
+        private val maxItems: Int
+            get() {
+                return widgetPrefs.widgetSize(appWidgetId).let { it.first * it.second }
+            }
+
+        data class WidgetFavorite(val title: String, val url: String, val bitmap: Bitmap?)
+        private val domains = mutableListOf<WidgetFavorite>()
+
+        override fun onCreate() {
+            inject(context)
+        }
+
+        override fun onDataSetChanged() {
```

Comment:
"onDataSetChanged() method, which gives you the opportunity to fetch any new data. Note that you can perform processing-intensive operations synchronously within the onDataSetChanged() callback."
https://developer.android.com/guide/topics/appwidgets/index.html#fresh